tstrait/simulate_effect_size.py

```diff
@@ -0,0 +1,170 @@
+        their genomic locations.
+        """
+        site_id = self.rng.choice(
+            range(self.ts.num_sites), size=self.num_causal, replace=False
+        )
+        site_id = np.sort(site_id)
+
+        return site_id
+
+    def _obtain_allele_count(self, tree, site):
+        """Obtain a dictionary of allele counts, and the ancestral state is not
+        included in the dictionary. Input is the tree sequence site (`ts.site(ID)`)
+        instead of site ID, as obtaining `ts.site(ID)` can be time consuming. The
+        ancestral state is not deleted if the ancestral state is the only allele
+        at that site.
+        """
+        counts = collections.Counter({site.ancestral_state: self.ts.num_samples})
+        for m in site.mutations:
+            current_state = site.ancestral_state
+            if m.parent != tskit.NULL:
+                current_state = self.ts.mutation(m.parent).derived_state
+            # Silent mutations do nothing
+            if current_state != m.derived_state:
+                num_samples = tree.num_samples(m.node)
+                counts[m.derived_state] += num_samples
+                counts[current_state] -= num_samples
+        del counts[site.ancestral_state]
+        counts = {x: y for x, y in counts.items() if y != 0}
+        if len(counts) == 0:
+            counts = {site.ancestral_state: self.ts.num_samples}
+        return counts
+
+    def _frequency_dependence(self, allele_freq):
+        if allele_freq == 0 or allele_freq == 1:
+            const = 0
+        else:
+            const = np.sqrt(pow(2 * allele_freq * (1 - allele_freq), self.alpha))
+        return const
+
+    def sim_causal_mutation(self):
+        """This method randomly chooses causal sites and the corresponding causal state
+        based on the `num_causal` input. Afterwards, effect size of each causal site
+        is simulated based on the trait model given by the `model` input.
+
+        :return: Returns a pandas dataframe that includes causal site ID, causal allele
+            and simulated effect size
+        :rtype: pandas.DataFrame
+        """
+        causal_site_array = self._choose_causal_site()
+        num_samples = self.ts.num_samples
+        tree = tskit.Tree(self.ts)
+
+        causal_state_array = np.zeros(self.num_causal, dtype=object)
+        beta_array = np.zeros(self.num_causal)
+
+        for i, single_id in enumerate(causal_site_array):
+            site = self.ts.site(single_id)
+            tree.seek(site.position)
+            counts = self._obtain_allele_count(tree, site)
+            causal_state = self.rng.choice(list(counts))
+            causal_state_array[i] = causal_state
+            allele_freq = counts[causal_state] / num_samples
+            beta = self.model.sim_effect_size(num_causal=self.num_causal, rng=self.rng)
+            beta *= self._frequency_dependence(allele_freq)
+            beta_array[i] = beta
+
+        effect_size_df = pd.DataFrame(
```

Comment:
This would be simpler:
```python
df = pd.DataFrame({"site_id"=causal_site_array, "causal_state": causal_state_array, 
      "effect_size": beta_array})
```
but maybe there's a performance penalty for doing it this way?